janusgraph-core/src/main/java/org/janusgraph/graphdb/query/graph/GraphCentricQueryBuilder.java

```diff
@@ -497,5 +407,122 @@ private static boolean coversAll(final MixedIndexType index, Condition<JanusGrap
         return match != null && indexInfo.supports(index, match, atom.getPredicate());
     }
 
+    private JointIndexQuery selectIndices(final Set<IndexType> indexCandidates, final MultiCondition<JanusGraphElement> conditions,
```

Comment:
I would suggest to move `selectIndices` method into a separate class. Something like `IndexSelectionService`